flink-runtime/src/main/java/org/apache/flink/runtime/state/v2/ttl/TtlListStateV2.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state.v2.ttl;
+
+import org.apache.flink.api.common.state.v2.StateFuture;
+import org.apache.flink.api.common.state.v2.StateIterator;
+import org.apache.flink.runtime.state.ttl.TtlStateContext;
+import org.apache.flink.runtime.state.ttl.TtlUtils;
+import org.apache.flink.runtime.state.ttl.TtlValue;
+import org.apache.flink.runtime.state.v2.adaptor.CompleteStateIterator;
+import org.apache.flink.runtime.state.v2.internal.InternalListState;
+import org.apache.flink.util.Preconditions;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * This class wraps list state with TTL logic.
+ *
+ * @param <K> The type of key the state is associated to
+ * @param <N> The type of the namespace
+ * @param <T> Type of the user entry value of state with TTL
+ */
+class TtlListStateV2<K, N, T>
+        extends AbstractTtlStateV2<K, N, T, TtlValue<T>, InternalListState<K, N, TtlValue<T>>>
+        implements InternalListState<K, N, T> {
+
+    protected TtlListStateV2(
+            TtlStateContext<InternalListState<K, N, TtlValue<T>>, T> ttlStateContext) {
+        super(ttlStateContext);
+    }
+
+    @Override
+    public StateFuture<Void> asyncUpdate(List<T> values) {
+        Preconditions.checkNotNull(values, "List of values to add cannot be null.");
+        return original.asyncUpdate(withTs(values));
+    }
+
+    @Override
+    public StateFuture<Void> asyncAddAll(List<T> values) {
+        Preconditions.checkNotNull(values, "List of values to add cannot be null.");
+        return original.asyncAddAll(withTs(values));
+    }
+
+    @Override
+    public StateFuture<StateIterator<T>> asyncGet() {
+        final List<T> result = new ArrayList<>();
+        AtomicBoolean encounteredNull = new AtomicBoolean(false);
+        AtomicBoolean anyUnexpired = new AtomicBoolean(false);
+        return original.asyncGet()
+                .thenAccept(
```

Comment:
How about using a iterator wrapper that cleans up the value during iteration instead of collect items in advance?

Reply:
Same for the iteration interface of `MapState`